### PR TITLE
Add dependency-graph information to `printPlan` output

### PR DESCRIPTION
**Pull request by hvr**
This improves the install-plan reporting in `cabal install -v2`'s output
to include reverse-dependencies, thereby giving valuable information why
a package was included in the install-plan to begin with.

Moreover, this also disables line-wrapping for the install-plan for
verbosity `verbose` and greater, to make the output easier to parse
for humans as well as computers

Example output for `cabal install --dry -v2 http-client`:

```
  In order, the following would be installed:
  base64-bytestring-1.0.0.1 (via: http-client-0.4.11.2) (new package)
  data-default-class-0.0.1 (via: http-client-0.4.11.2 cookie-0.4.1.5) (new package)
  mtl-2.2.1 (via: exceptions-0.8.0.2 parsec-3.1.9) (new package)
  network-2.6.0.2 (via: http-client-0.4.11.2 streaming-commons-0.1.12) (new package)
  old-locale-1.0.0.7 (via: cookie-0.4.1.5) (new package)
  random-1.1 (via: http-client-0.4.11.2 streaming-commons-0.1.12) (new package)
  stm-2.4.4 (via: streaming-commons-0.1.12 exceptions-0.8.0.2) (new package)
  text-1.2.0.4 -integer-simple (via: http-client-0.4.11.2 streaming-commons-0.1.12 parsec-3.1.9 mime-types-0.1.0.6 http-types-0.8.6 case-insensitive-1.2.0.4 hashable-1.2.3.2 cookie-0.4.1.5 blaze-builder-0.4.0.1) (new package)
  blaze-builder-0.4.0.1 (via: http-client-0.4.11.2 streaming-commons-0.1.12 http-types-0.8.6 cookie-0.4.1.5) (new package)
  cookie-0.4.1.5 (via: http-client-0.4.11.2) (new package)
  hashable-1.2.3.2 (via: case-insensitive-1.2.0.4) (new package)
  case-insensitive-1.2.0.4 (via: http-client-0.4.11.2 http-types-0.8.6) (new package)
  http-types-0.8.6 (via: http-client-0.4.11.2) (new package)
  mime-types-0.1.0.6 (via: http-client-0.4.11.2) (new package)
  parsec-3.1.9 (via: network-uri-2.6.0.1) (new package)
  network-uri-2.6.0.1 (via: http-client-0.4.11.2) (new package)
  transformers-compat-0.4.0.4 (via: exceptions-0.8.0.2) (new package)
  exceptions-0.8.0.2 (via: http-client-0.4.11.2) (new package)
  zlib-0.6.1.0 (via: streaming-commons-0.1.12) (new package)
  streaming-commons-0.1.12 (via: http-client-0.4.11.2) (new package)
  http-client-0.4.11.2 (new package)
```

/cc @kosmikus @dcoutts

### CODE REVIEW

**Comment by hvr** [comment]
NB: this is against the 1.22 branch, as that's where I needed it for; it should be trivial to port to `master`; I filed this PR mostly to start the discussion/bikeshedding

**Comment by kosmikus** [comment]
In principle, this is certainly a good idea. If a human is the intended target, I'd actually prefer if we could print this as a tree, rather than explicitly listing the reverse deps. But that might defeat your purpose.

**Comment by 23Skidoo** [comment]
+1 on printing this as a tree, would make it easier to read.

**Comment by hvr** [comment]
Tbh, I'm not sure how to represent the dependency DAG accurately as a tree... can you elaborate how "diamonds" would be represented?

File: cabal-install/Distribution/Client/Install.hs [hunk 1]
```diff
@@ -582,7 +583,7 @@ printPlan :: Bool -- is dry run
 printPlan dryRun verbosity plan sourcePkgDb = case plan of
   []   -> return ()
   pkgs
-    | verbosity >= Verbosity.verbose -> notice verbosity $ unlines $
+    | verbosity >= Verbosity.verbose -> putStr $ unlines $
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/Install.hs]: Why this change?

Comment (hvr) [inline cabal-install/Distribution/Client/Install.hs]: It was mentioned in the commit msg... it's so that the output becomes easier to parse, as the line-wrapping done by `notice` makes it much harder to write a parser for the output. Moreover, even when looking myself at the out, I consider the line-wrapping rather distracting and making the output actually less readable...

Comment (23Skidoo) [inline cabal-install/Distribution/Client/Install.hs]: Thanks, I should have read the commit message.

**Comment by 23Skidoo** [comment]
I think that this can be merged right now. Improvements, such as printing as a tree, can come later.

**Comment by hvr** [comment]
@23Skidoo do you want me to send in a PR for `master` too?

PS: what about optionally (e.g. `--dump-dot-install-plan <outfile>`) dumping a graphviz style (c.f. `ghc-pkg dot` output) of the *full* forward deps? IMHO we can't improve the output much more w/o implementing some text-based graph-rendering...

**Comment by 23Skidoo** [comment]
Sure, please do.

> PS: what about optionally (e.g. `--dump-dot-install-plan <outfile>`) dumping a graphviz style (c.f. `ghc-pkg dot` output) of the *full* forward deps? 

That'd be nice to have too.
